### PR TITLE
fluent-bit shared object go plugin

### PR DESCRIPTION
**Pull request by cosmo0920**
**What this PR does / why we need it**:

The first step to merge https://github.com/cosmo0920/fluent-bit-go-loki.

**Which issue(s) this PR fixes**:
No fixes. This is the first task of https://github.com/grafana/loki/issues/770.

**Special notes for your reviewer**:

I think that we would have to create `fluent-bit/fluent-bit-go-loki` folder instead of `cmd/fluent-bit`.
Because fluent-bit-go-loki Golang plugin uses shared object format.
It is not executable.

### Building steps for fluent-bit-go-loki

fluent-bit-go-loki can build with:

```bash
$ (cd fluent-bit/fluent-bit-go-loki && go build -buildmode=c-shared -o out_loki.so .)
```

Or, make with added Makefile(building and testing):

```bash
$ (cd fluent-bit/fluent-bit-go-loki && make)
```

### Testing for fluent-bit-go-loki

fluent-bit-go-loki test can execute with:

```bash
$ (cd fluent-bit/fluent-bit-go-loki && go test -cover -race -coverprofile=coverage.txt -covermode=atomic)
```

Or, make with added Makefile:

```bash
$ (cd fluent-bit/fluent-bit-go-loki && make test)
```

**Checklist**
- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

File: fluent-bit/fluent-bit-go-loki/Makefile [hunk 1]
```diff
@@ -0,0 +1,14 @@
+all: test
```
Comment (sh0rez) [inline fluent-bit/fluent-bit-go-loki/Makefile]: Any reason this is a separate file? Could it be integrated into the main Makefile?

File: fluent-bit/fluent-bit-go-loki/README.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
+# fluent-bit loki output plugin
+
+This plugin works with fluent-bit's go plugin interface. You can use fluent-bit loki to ship logs into grafana datasource with loki.
```
Comment (sh0rez) [inline fluent-bit/fluent-bit-go-loki/README.md]: Shouldn't this be 
`It allows to fluent-bit to ship logs into Loki, for use with e.g. Grafana?`

File: fluent-bit/fluent-bit-go-loki/loki.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+	batchSizeValue, err := strconv.Atoi(batchSize)
+	if err != nil || batchSize == "" {
+		batchSizeValue = 10
+	}
+	lc.batchSize = batchSizeValue * 1024
+
+	var labelValues labelSetJSON
+	if labels == "" {
+		labels = `
```
Comment (sh0rez) [inline fluent-bit/fluent-bit-go-loki/loki.go]: Are the `\n` before and after required?

File: fluent-bit/fluent-bit-go-loki/out_loki.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+	os.Exit(code)
+}
+
+func (p *fluentPlugin) HandleLine(ls model.LabelSet, timestamp time.Time, line string) error {
+	return loki.Handle(ls, timestamp, line)
+}
+
+//export FLBPluginRegister
+func FLBPluginRegister(ctx unsafe.Pointer) int {
+	return output.FLBPluginRegister(ctx, "loki", "Loki Output plugin written in GO!")
```
Comment (sh0rez) [inline fluent-bit/fluent-bit-go-loki/out_loki.go]: Maybe this should be something more descriptive, e.g. `Output plugin for Grafana Loki` or `Ship logs to Grafana Loki`

File: fluent-bit/fluent-bit-go-loki/Makefile (hunk 1 above)
Comment (cosmo0920) [reply fluent-bit/fluent-bit-go-loki/Makefile]: Oh. Your comments is reasonable.
Hm, how about remove this file in the first step?

I'm struggling to understand Loki Makefile structure. 😅

Comment (sh0rez) [reply fluent-bit/fluent-bit-go-loki/Makefile]: Okay, no problem:

`all`: remove
`fast`: add to current makefile, rename something like `fluent-bit`
`test`: remove, already covered by root `make test`
`dep`: remove
`clean`: add the command to the current `make clean`, note to prefix the wildcards with the correct directory

Comment (cosmo0920) [reply fluent-bit/fluent-bit-go-loki/Makefile]: Wow. It's super fast advise!

I've modified the current makefile as follows:

`all`: removed
`fast`: renamed to `fluent-bit` and use `go build -buildmode=c-shared -o out_loki.so .` command
`test`: removed
`dep`: removed
`clean`: Use concrete file name. Without wildcards.

File: fluent-bit/fluent-bit-go-loki/loki.go (hunk 3 above)
Comment (cosmo0920) [reply fluent-bit/fluent-bit-go-loki/loki.go]: Oh, good catch. They needn't.

File: fluent-bit/fluent-bit-go-loki/Makefile (hunk 1 above)
Comment (sh0rez) [reply fluent-bit/fluent-bit-go-loki/Makefile]: Nice! Could you integrate those targets into the `/Makefile`?

**Comment by sh0rez** [comment]
Regarding the placement of the fluentd / fluent-bit plugins ... Couldn't we place them in a common `fluent` folder? Something like

```tree
.
├── fluent
│   ├── bit
│   └── d
```

/cc @cyriltovena @slim-bean

File: fluent-bit/fluent-bit-go-loki/Makefile (hunk 1 above)
Comment (cosmo0920) [reply fluent-bit/fluent-bit-go-loki/Makefile]: Sorry, I'm still not to be able to integrate them into the `/Makefile`...

Comment (sh0rez) [reply fluent-bit/fluent-bit-go-loki/Makefile]: Okay, no problem. May I help out?

Comment (sh0rez) [reply fluent-bit/fluent-bit-go-loki/Makefile]: Okay, below is a patch that includes the changes.

Save to a file, apply using `git am < filename`.

```patch
From dfc82f917839704a367b7b98c887fd3295ec57aa Mon Sep 17 00:00:00 2001
From: sh0rez <me@shorez.de>
Date: Mon, 5 Aug 2019 12:12:46 +0200
Subject: [PATCH] refactor(fluent-bit): move make targets into root Makefile

---
 Makefile                               | 7 +++++++
 fluent-bit/fluent-bit-go-loki/Makefile | 5 -----
 2 files changed, 7 insertions(+), 5 deletions(-)
 delete mode 100644 fluent-bit/fluent-bit-go-loki/Makefile

diff --git a/Makefile b/Makefile
index c05b5e4b..44074406 100644
--- a/Makefile
+++ b/Makefile
@@ -210,6 +210,7 @@ clean:
 	rm -rf .cache
 	rm -rf cmd/docker-driver/rootfs
 	rm -rf dist/
+	rm -rf fluent-bit/fluent-bit-go-loki/out_loki.*
 	go clean ./...
 
 #########
@@ -333,6 +334,12 @@ docker-driver-clean:
 	-docker plugin rm grafana/loki-docker-driver:$(IMAGE_TAG)
 	rm -rf cmd/docker-driver/rootfs
 
+#####################
+# fluent-bit plugin #
+#####################
+fluent-bit-plugin:
+	go build -buildmode=c-shared -o fluent-bit/fluent-bit-go-loki/out_loki.so ./fluent-bit/fluent-bit-go-loki
+
 ########################
 # Bigtable Backup Tool #
 ########################
diff --git a/fluent-bit/fluent-bit-go-loki/Makefile b/fluent-bit/fluent-bit-go-loki/Makefile
deleted file mode 100644
index b3a25f36..00000000
--- a/fluent-bit/fluent-bit-go-loki/Makefile
+++ /dev/null
@@ -1,5 +0,0 @@
-fluent-bit:
-	go build -buildmode=c-shared -o out_loki.so .
-
-clean:
-	rm -rf out_loki.so out_loki.h
-- 
2.17.1

```

File: fluent-bit/fluent-bit-go-loki/loki.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
+package main
+
+import "github.com/cortexproject/cortex/pkg/util/flagext"
+import "github.com/prometheus/common/model"
```
Comment (cyriltovena) [inline fluent-bit/fluent-bit-go-loki/loki.go]: Let's organize import all together like this:
```
import (
native..

others...
)
```

File: fluent-bit/fluent-bit-go-loki/Makefile [hunk 6]
```diff
@@ -0,0 +1,5 @@
+fluent-bit:
```
Comment (cyriltovena) [inline fluent-bit/fluent-bit-go-loki/Makefile]: If you want to move this into our Makefile later, let's create an issue and we will get back at it.

File: fluent-bit/fluent-bit-go-loki/Makefile (hunk 1 above)
Comment (cosmo0920) [reply fluent-bit/fluent-bit-go-loki/Makefile]: Thanks! I've confirmed your patch is working! 😃

File: fluent-bit/fluent-bit-go-loki/Makefile (hunk 6 above)
Comment (cosmo0920) [reply fluent-bit/fluent-bit-go-loki/Makefile]: I've imported @sh0rez patch:
https://github.com/grafana/loki/pull/847/commits/16af27376b8a2fe190b3f507fc59cfca78a9449f

**Comment by cyriltovena** [comment]
Can you confirm the new code in `fluent-bit` folder is linted by the linter ?

**Comment by cosmo0920** [comment]
golangci-lint complains the following warning:

```log
% GOGC=20 golangci-lint run
WARN [runner/golint] Golint: can't lint 5 files: no file name for file &{Doc:<nil> Package:21595946 Name:main Decls:[0xc0287bd040 0xc0287bd180 0xc0287bd2c0 0xc0287bfb60] Scope:scope 0xc028786dc0 {
	func getLokiConfig
	type lokiConfig
	type labelSetJSON
}
 Imports:[0xc0287bf1a0 0xc0287bf1d0 0xc0287bf200 0xc0287bf230 0xc0287bf260 0xc0287bf290] Unresolved:[flagext time int model string string string string string string error flagext nil nil fmt strconv nil time time strconv nil json byte nil nil fmt make model model model nil] Comments:[]} 
```

But I have no idea what it means. :sweat:

**Comment by sh0rez** [comment]
> Can you confirm the new code in `fluent-bit` folder is linted by the linter ?

It is
https://github.com/grafana/loki/blob/a22be1785933830603648b2b8ab466e7e8ecadac/Makefile#L191-L192
This lints everything it can find

**Comment by cyriltovena** [comment]
It should be the import issue passed and I think there is more but the linter seems to be skipping.

Let me have quick look.

**Comment by cyriltovena** [comment]
You can remove the license file we have the same.

File: fluent-bit/fluent-bit-go-loki/out_loki.go [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+		case output.FLBTime:
+			timestamp = ts.(output.FLBTime).Time
+		case uint64:
+			timestamp = time.Unix(int64(t), 0)
+		default:
+			fmt.Print("timestamp isn't known format. Use current time.")
+			timestamp = time.Now()
+		}
+
+		line, err := createJSON(record)
```
Comment (cyriltovena) [inline fluent-bit/fluent-bit-go-loki/out_loki.go]: There is something wrong here Loki is supposed to received only the the actual text as line. It seems that you are also embedding metadata in it using json.

I think you should translate records into labels (except for the line part) and merge it with the global labels values from the configuration.

Most people will use that fluentbit with the kubernetes plugin so discovery will be done that way and the record will contain information like pod_name, container_id ..... and those should not be part of the line sent to Loki.

File: fluent-bit/fluent-bit-go-loki/out_loki.go [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+		if err != nil {
+			fmt.Printf("error creating message for Grafana Loki: %v", err)
+			continue
+		}
+
+		err = plugin.HandleLine(ls, timestamp, line)
+		if err != nil {
+			fmt.Printf("error sending message for Grafana Loki: %v", err)
+			return output.FLB_RETRY
```
Comment (cyriltovena) [inline fluent-bit/fluent-bit-go-loki/out_loki.go]: There is already a retry mechanism inside our client. I would just return `output.FLB_ERROR`.

File: fluent-bit/fluent-bit-go-loki/version.go [hunk 9]
```diff
@@ -0,0 +1,3 @@
+package main
+
+const Version = "0.1.3"
```
Comment (cyriltovena) [inline fluent-bit/fluent-bit-go-loki/version.go]: Would be nice if you could tag this the same way we tag Loki binaries. (eg see `GO_FLAGS` in our makefile.)

File: fluent-bit/fluent-bit-go-loki/out_loki.go (hunk 7 above)
Comment (cyriltovena) [reply fluent-bit/fluent-bit-go-loki/out_loki.go]: see https://fluentbit.io/documentation/0.12/filter/kubernetes.html

Comment (cyriltovena) [reply fluent-bit/fluent-bit-go-loki/out_loki.go]: check our fluentd configuration https://github.com/grafana/loki/tree/master/fluentd/fluent-plugin-grafana-loki#output-format

That should give you a sense of what is required.

File: fluent-bit/fluent-bit-go-loki/out_loki.go [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+
+		// Get timestamp
+		var timestamp time.Time
+		switch t := ts.(type) {
+		case output.FLBTime:
+			timestamp = ts.(output.FLBTime).Time
+		case uint64:
+			timestamp = time.Unix(int64(t), 0)
+		default:
+			fmt.Print("timestamp isn't known format. Use current time.")
```
[the remark this answers is not in the document]
Comment (cosmo0920) [reply fluent-bit/fluent-bit-go-loki/out_loki.go]: I see. I'll try it.
But when using your logger library in this plugin, we shall add a new configuration to specify log level. 
Is it acceptable?

File: fluent-bit/fluent-bit-go-loki/out_loki.go (hunk 7 above)
Comment (cosmo0920) [reply fluent-bit/fluent-bit-go-loki/out_loki.go]: fluent-bit has loosely coupled plugin ecosystem and they loves to single functional plugin.
Users should consider to filter record with another plugin (such as filter grep) in their fluent-bit data pipeline:
https://docs.fluentbit.io/manual/v/1.2/filter/grep

IMO, I don't want to add record manipulation functionality in this plugin.
But it is OK to add.

File: fluent-bit/fluent-bit-go-loki/out_loki.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (sh0rez) [reply fluent-bit/fluent-bit-go-loki/out_loki.go]: Can't you get the log level from fluent-bit? It should have its own

[the remark this answers is not in the document]
Comment (cosmo0920) [reply fluent-bit/fluent-bit-go-loki/out_loki.go]: > Can't you get the log level from fluent-bit? 

No, it cannot.
fluent-bit-go shared object plugin does not have handler for fluent-bit's system log level.

> It should have its own

:+1:

[the remark this answers is not in the document]
Comment (sh0rez) [reply fluent-bit/fluent-bit-go-loki/out_loki.go]: Okay, then you need to use your own one

[the remark this answers is not in the document]
Comment (cyriltovena) [reply fluent-bit/fluent-bit-go-loki/out_loki.go]: yeah you could have a env variable for that.

File: fluent-bit/fluent-bit-go-loki/version.go (hunk 9 above)
Comment (cosmo0920) [reply fluent-bit/fluent-bit-go-loki/version.go]: You means that using build-time variable to specify plugin version, right?

But other option which is `-a -tags netgo -installsuffix netgo --ldflags '-extldflags "-static"'` is unusable for building plugin because it is not executables but shared object:

```log
% LANG=C go build -a -tags netgo -installsuffix netgo --ldflags '-extldflags "-static"' -buildmode=c-shared -o out_loki.so .
# github.com/cosmo0920/fluent-bit-go-loki
/home/hhatake/.goenv/versions/1.12.7/pkg/tool/linux_amd64/link: running gcc failed: exit status 1
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/6/crtbeginT.o: relocation R_X86_64_32 against hidden symbol `__TMC_END__' can not be used when making a shared object
/usr/bin/ld: final link failed: Nonrepresentable section on output
collect2: error: ld returned 1 exit status
```

Comment (cosmo0920) [reply fluent-bit/fluent-bit-go-loki/version.go]: I've fixed with https://github.com/grafana/loki/pull/847/commits/c18bc38813c2c1d1e8996fde1da1f0e20c67222e.

**Comment by cyriltovena** [comment]
@cosmo0920 I'm on vacation for 2 weeks, will review asap when I'm back.

**Comment by cosmo0920** [comment]
Yep, Thank you for notifying it.
I'm also on vacation for a week and a half.

**Comment by cyriltovena** [comment]
I'm taking another look !

File: fluent-bit/fluent-bit-go-loki/loki.go [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+	if err != nil || batchSize == "" {
+		batchSizeValue = 100 * 1024
+	}
+	lc.batchSize = batchSizeValue
+
+	if labels == "" {
+		labels = `{job="fluent-bit"}`
+	}
+
+	ls, err := logql.ParseExpr(labels)
```
Comment (cyriltovena) [inline fluent-bit/fluent-bit-go-loki/loki.go]: you can use the new https://github.com/grafana/loki/blob/master/pkg/logql/parser.go#L47 `logql.ParseMatchers`

File: fluent-bit/fluent-bit-go-loki/loki_test.go [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGetLokiConfig(t *testing.T) {
+	c, err := getLokiConfig("", "", "", "", "", "")
```
Comment (cyriltovena) [inline fluent-bit/fluent-bit-go-loki/loki_test.go]: Table tests are perfectly suited for this. Let's keep this for another PR.

File: fluent-bit/fluent-bit-go-loki/out_loki.go [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+		if ret != 0 {
+			break
+		}
+
+		// Get timestamp
+		var timestamp time.Time
+		switch t := ts.(type) {
+		case output.FLBTime:
+			timestamp = ts.(output.FLBTime).Time
+		case uint64:
```
Comment (cyriltovena) [inline fluent-bit/fluent-bit-go-loki/out_loki.go]: is this case required ? How do you know this can happens ?

File: fluent-bit/fluent-bit-go-loki/out_loki.go (hunk 7 above)
Comment (cyriltovena) [reply fluent-bit/fluent-bit-go-loki/out_loki.go]: I think we should support the same config as fluentd plugin except `drop_single_key`. You're very close you just need those 3:

remove_keys: (default: nil) comma separated list of needless record keys to remove. All other keys will be placed into the log line

label_keys: (default: "job,instance") comma separated list of keys to use as stream labels. All other keys will be placed into the log line

line_format: format to use when flattening the record to a log line. Valid values are "json" or "key_value". If set to "json" the log line sent to Loki will be the fluentd record (excluding any keys extracted out as labels) dumped as json. If set to "key_value", the log line will be each item in the record concatenated together (separated by a single space) in the format <key>=<value>.

**Comment by JensErat** [comment]
When putting together a proof of concept towards Loki, we also had to extend the fluent-bit output plugin due to some issues. I will definitely extract and provide a new PR for some useful parts as soon as this PR is merged (especially added quite some testing code because debugging edge cases is a hassle with CGO shared objects).

Mostly, our issues boil down to fluent-bit records not necessarily being flat, but can be a tree-like structure. For example, the fluent-bit Kubernetes filter plugin will add meta information from Kubernetes (pod, namespace, ...) within a `kubernetes` "folder", and if you enable JSON parsing, whole nested JSON trees can be passed as record (think of [Kubernetes audit logs](https://medium.com/@noqcks/kubernetes-audit-logging-introduction-464a34a53f6c) that sometimes contain three or four times nested records).

This resulted in two issues for us:

- Extracting stream labels requires a non-trivial configuration. We went for JSON like this:

    ```json
    {
      "TENANT": "tenant",
      "HOSTNAME": "hostname",
      "kubernetes": {
        "container_name": "container",
        "pod_name": "pod",
        "namespace_name": "namespace"
      },
      "BOOT_ID": "_drop",
      "CONTAINER_ID": "_drop",
      "CONTAINER_ID_FULL": "_drop",
      "CONTAINER_TAG": "_drop",
      "SYSTEMD_INVOCATION_ID": "_drop",
      "SELINUX_CONTEXT": "_drop",
      "CONTAINER_NAME": "_drop",
      "MACHINE_ID": "_drop"
    }
    ```

    The semantics of this is the key being a match on the fluent-bit record label, and the value either a nested sub-match (`container_name` from `kubernetes` dictionary), the target label (which is flat in Loki) or a specific drop annotation. This is easily implementable using a recursive function (we have partially test-backed prototype code). This is pretty much a tree-aware alternative to the fluentd-plugin `remove_keys` and `label_keys` configuration. The example above keeps `TENANT` and `HOSTNAME` in their lower-case variants, and renames `kubernetes.pod_name` etc.  to simple `pod` labels (we see them in queries and write them in Grafana all the time, so rather keep them short).

    Surely this is also implementable as a more simple data structure, but going for a well-defined JSON-based configuration resolved lots of issues around escaping of special characters.

- The `[]byte`/base64 conversion has also to be applied to keys and values beyond the outermost record layer. We wondered about lots of base64-encoded text until we realized this is based on nested labels.

I'm currently doing some minor cleanups and will push the code as-is to a fork. We would be glad if we can find some agreements on how to handle nested labels with both of the issues they bring -- and we're ready to provide code to get them running, we just have to align on how to do it properly.

**Comment by cyriltovena** [comment]
out of curiosity do you think the same issue also happen in fluentd plugin ? For promtail we have pipeline stage to parse json however we don't yet support `drop`. 

This looks like a nice solution but we need to make sure we use the same config across the board unless there is a good argument against.

/cc @slim-bean @rfratto WDYT ?

**Comment by JensErat** [comment]
> out of curiosity do you think the same issue also happen in fluentd plugin ? For promtail we have pipeline stage to parse json however we don't yet support `drop`.

It depends. I would expect fluentd can also handle nested records. After all, the whole protobuf forward protocol originated in fluentd as far as I know, not in fluent-bit.

On the other hand, the `[]byte`/base64 issue is probably specific to the Golang protobuf unmarshaller and json marshaller, respective the golang string conversion in general. I guess you wouldn't have this issuer in either Ruby nor C.

> This looks like a nice solution but we need to make sure we use the same config across the board unless there is a good argument against.

Definitely. I don't entirely like the configuration syntax I proposed -- I just couldn't come up with something better yet (and it works pretty well). So if anybody comes up with better thoughts... To me it looks like fluentd would support "structured" configuration values anyway: https://docs.fluentd.org/configuration/config-file#format-tips (so I'm not sure whether one should go for supporting JSON over there or better rely on what would be fluentd native).

File: fluent-bit/fluent-bit-go-loki/out_loki.go (hunk 13 above)
Comment (cosmo0920) [reply fluent-bit/fluent-bit-go-loki/out_loki.go]: This should be required because Fluentd events can contain two time formats which are `EventTime` and Unix epoch.

`EventTime` will be translated as `output.FLBTime` type.
Unix epoch will be translated as uint64 type.

ref: https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1#message-modes
ref: https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1#eventtime-ext-format

**Comment by cyriltovena** [comment]
@JensErat I think this would be a great additions to this PR once merged.

**Comment by cyriltovena** [comment]
I'm going to refactor this a bit and get it merged.

**Comment by cyriltovena** [comment]
@joe-elliott this is ready for review, I'm going to add a section on how to run using docker in the readme once we got this merged, currently there is no image pushed yet.

I refactored the code to be easier to maintain in the future, and also added docker integration.

@JensErat I would love a PR on top of this one allowing to parse nested records.

**Comment by cyriltovena** [comment]
I could not help myself and fix the label nesting problem. @JensErat see my approach.

I'll write more doc and we should be good to merge.

File: .drone/drone.yml [hunk 14]
```diff
@@ -139,6 +139,40 @@ steps:
       include:
       - refs/heads/master
       - refs/tags/v*
+- depends_on:
```
Comment (joe-elliott) [inline .drone/drone.yml]: It appears to me that drone.yml is generated from drone.jsonnet via the Makefile.  Should these changes be in drone.jsonnet?

File: fluent-bit/fluent-bit-go-loki/Dockerfile [hunk 15]
```diff
@@ -0,0 +1,9 @@
+FROM golang:1.12 as build
```
Comment (joe-elliott) [inline fluent-bit/fluent-bit-go-loki/Dockerfile]: should this go under the `./cmd` subtree?

File: fluent-bit/fluent-bit-go-loki/kubernetes-example.yaml [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+        k8s-app: fluent-bit-logging
+        version: v1
+        kubernetes.io/cluster-service: "true"
+    spec:
+      containers:
+      - name: fluent-bit
+        image: grafana/fluent-bit-plugin-loki:latest
+        env:
+          - name:  LOKI_URL
+            value: http://host.docker.internal:3100/api/prom/push
```
Comment (joe-elliott) [inline fluent-bit/fluent-bit-go-loki/kubernetes-example.yaml]: remove in favor of new path?

File: tools/fluent-bit.conf [hunk 17]
```diff
@@ -0,0 +1,14 @@
+[INPUT]
+    Name        tail
+    Path        /var/log/*.log
+[Output]
+    Name loki
+    Match *
+    Url http://host.docker.internal:3100/api/prom/push
```
Comment (joe-elliott) [inline tools/fluent-bit.conf]: remove in favor of new path?

File: .drone/drone.yml (hunk 14 above)
Comment (cyriltovena) [reply .drone/drone.yml]: really ?

File: fluent-bit/fluent-bit-go-loki/kubernetes-example.yaml (hunk 16 above)
Comment (cyriltovena) [reply fluent-bit/fluent-bit-go-loki/kubernetes-example.yaml]: This file is gone, I've pushed a helm chart, last bit is more documentation for how to run inside ocker/k8s or linux.

File: tools/fluent-bit.conf (hunk 17 above)
Comment (cyriltovena) [reply tools/fluent-bit.conf]: yes I rebase to master also to get the latest endpoints.

File: fluent-bit/fluent-bit-go-loki/Dockerfile (hunk 15 above)
Comment (cyriltovena) [reply fluent-bit/fluent-bit-go-loki/Dockerfile]: I'm not against the idea 🤷‍♂

File: .drone/drone.yml (hunk 14 above)
Comment (sh0rez) [reply .drone/drone.yml]: Yes, definitely! The `drone.yml` should never be edited by hand

File: fluent-bit/fluent-bit-go-loki/Dockerfile (hunk 15 above)
Comment (cyriltovena) [reply fluent-bit/fluent-bit-go-loki/Dockerfile]: I moved it there.

File: .drone/drone.yml (hunk 14 above)
Comment (cyriltovena) [reply .drone/drone.yml]: doh !

**Comment by JensErat** [comment]
Thank you for the great work -- I didn't get round checking the code before now again. I really like the solution with separate relabel- and drop configuration though. At first I tried rebasing my changes against the current HEAD (as I somehow was not aware you reimplemented the small recursive function already) -- pretty much puzzled my because you came up with nearly 100% identical code but on the "wrong side" of the diffs until I realized what was going on... :)

I ported some of my edge case tests I wrote while drafting my code:

https://github.com/grafana/loki/pull/1096
